File: 백준/Gold/16472. 고냥이/고냥이.py

n = int(input())
arr = list(input())

ans = start = 0
dic = {}
for end in range(len(arr)):
    dic.setdefault(arr[end], 0)
    dic[arr[end]] += 1
    while len(dic) > n: 
        dic[arr[start]] -=1
        if dic[arr[start]] == 0:
            del dic[arr[start]]
        start+=1
    ans = max(ans, end-start+1)
print(ans)

File: 백준/Gold/16472. 고냥이/고냥이.java
import java.util.Scanner;

public class Main {

  public static void main(String[] args) {
    Scanner sc = new Scanner(System.in);
    int n = sc.nextInt();
    char[] arr = sc.next().toCharArray();
    int[] alp = new int[27];
    int ans = 0;
    int nextIdx = 0;
    for (int i = 0; i < arr.length; i++) {
      while (nextIdx < arr.length ){
        alp[arr[nextIdx++] - 'a']++;
        if (countAlp(alp) > n) break;
        ans = Math.max(ans, nextIdx - i);
//        System.out.println(i +" " + n extIdx);
      }
      alp[arr[i] - 'a'] --;
    }
    System.out.println(ans);
  }

  private static int countAlp(int[] alp) {
    int cnt =0;
    for (int i = 0; i < 26; i++) {
      if (alp[i] > 0) cnt++;
    }
    return cnt;
  }
}
